controllers/kustomization_controller.go

```diff
@@ -123,278 +126,272 @@ func (r *KustomizationReconciler) SetupWithManager(mgr ctrl.Manager, opts Kustom
 	snapshot, err := r.build(ctx, kustomization, checksum, dirPath)
 	if err != nil {
-		return kustomizev1.KustomizationNotReady(
-			kustomization,
-			source.GetArtifact().Revision,
-			kustomizev1.BuildFailedReason,
-			err.Error(),
-		), err
+		conditions.MarkFalse(kustomization, meta.ReadyCondition, kustomizev1.BuildFailedReason, err.Error())
+		return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, err
 	}
 
 	// dry-run apply
 	err = r.validate(ctx, kustomization, impersonation, dirPath)
 	if err != nil {
-		return kustomizev1.KustomizationNotReady(
-			kustomization,
-			source.GetArtifact().Revision,
-			kustomizev1.ValidationFailedReason,
-			err.Error(),
-		), err
+		conditions.MarkFalse(kustomization, meta.ReadyCondition, kustomizev1.ValidationFailedReason, err.Error())
+		return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, err
 	}
 
 	// apply
-	changeSet, err := r.applyWithRetry(ctx, kustomization, impersonation, source.GetArtifact().Revision, dirPath, 5*time.Second)
+	changeSet, err := r.applyWithRetry(ctx, *kustomization, impersonation, artifact.Revision, dirPath, 5*time.Second)
 	if err != nil {
-		return kustomizev1.KustomizationNotReady(
-			kustomization,
-			source.GetArtifact().Revision,
-			meta.ReconciliationFailedReason,
-			err.Error(),
-		), err
+		conditions.MarkFalse(kustomization, meta.ReadyCondition, meta.ReconciliationFailedReason, err.Error())
+		return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, err
 	}
 
 	// prune
 	err = r.prune(ctx, kubeClient, kustomization, checksum)
 	if err != nil {
-		return kustomizev1.KustomizationNotReady(
-			kustomization,
-			source.GetArtifact().Revision,
-			kustomizev1.PruneFailedReason,
-			err.Error(),
-		), err
+		conditions.MarkFalse(kustomization, meta.ReadyCondition, kustomizev1.PruneFailedReason, err.Error())
+		return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, err
 	}
 
 	// health assessment
-	err = r.checkHealth(ctx, statusPoller, kustomization, source.GetArtifact().Revision, changeSet != "")
+	err = r.checkHealth(ctx, statusPoller, kustomization, artifact.Revision, changeSet != "")
 	if err != nil {
-		return kustomizev1.KustomizationNotReadySnapshot(
-			kustomization,
-			snapshot,
-			source.GetArtifact().Revision,
-			kustomizev1.HealthCheckFailedReason,
-			err.Error(),
-		), err
-	}
-
-	return kustomizev1.KustomizationReady(
-		kustomization,
-		snapshot,
-		source.GetArtifact().Revision,
-		meta.ReconciliationSucceededReason,
-		"Applied revision: "+source.GetArtifact().Revision,
-	), nil
+		conditions.MarkFalse(kustomization, meta.ReadyCondition, kustomizev1.HealthCheckFailedReason, err.Error())
+		conditions.MarkFalse(kustomization, kustomizev1.HealthyCondition, kustomizev1.HealthCheckFailedReason, err.Error())
+		return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, err
+	}
+
+	kustomization.Status.LastAppliedRevision = artifact.Revision
+	kustomization.Status.Snapshot = snapshot
```

Comment:
I would set this as soon as it can be set, and I think that might be a bit earlier.

@stefanprodan: at what time becomes the previous snapshot irrelevant?

Reply:
We have to set the snapshot right after prune.